# DISCOVERING AND ASSIGNING TOPICS FOR ADVOCACY CAMPAIGN PETITION MAILINGS USING LATENT DIRICHLET ALLOCATION (LDA)

### In this phase of my project I use machine learning models to cluster SumOfUs mailings into issue topics. Then, for each topic, I assign a percentage to each campaign, based on the probability that topic relates to that campaign.  The algorthim I am using is called Latent Dirichlet Allocation (LDA).  

### The LDA algoritm will process the text of 1399 SumOfUs advocacy mailings and discover the top 8 issues, or topics, in SumOfUs campaigns.  The LDA will then go back through each mailing and assign it a percentage for each of the 8 topics.  Some campaigns pertain only to one topic, while others might relate to several topics in differing degreees.

### After I assign the percentages to each campaign, I will go on to the next phase, where I will combine this information with other data about the campaign, such as virality, regional distribution, and the very early behavior of new joiners on that campaign.  All of this will go into a different model, called a logistic regression, that will help me to predict the donation propensity of a new cohort only a week after joining the list.

In [1]:
#import the first Python modules I will be using in the project
import pandas as pd  #for working with data in table form within Python
import numpy as np   #important math and logic functions

### First, I read in the CSV file that I prepared in the Data Wrangling phase.  The file contains the page id, the name of each mailing tag, and the full HTML in each mailing. Since each mailing can have more than one tag, most mailings  have multiple rows in the file.

In [2]:
camp_txt = pd.read_csv('../capstone/page_mailing_selected.csv', encoding = "ISO-8859-1")  #import CSV as a Pandas table
camp_txt.tail(10)  #displays the last 10 rows of the table

page_id                               tag_name  \
2092    16154                           #Environment   
2093    16148                           #Environment   
2094    16160                                   bees   
2095    16160                           #Environment   
2096    16416                           #Environment   
2097    16458                        #Workers_Rights   
2098    16479                        #Workers_Rights   
2099    16736  #Privatization_and_Political_Meddling   
2100    16940                            oil company   
2101    16940                           #Environment   

                                                   html  
2092  <div style=width: 320px; float: right;>\r\n<ta...  
2093  <div style=width: 320px; float: right;>\r\n<ta...  
2094  <div style=width: 320px; float: right;>\r\n<ta...  
2095  <div style=width: 320px; float: right;>\r\n<ta...  
2096  <p>Dear {{ user.first_name|capfirst|default:Fr...  
2097  <div style=width: 320px; float: right;>\r\n<ta...  
2098  <div style=width: 320px; float: right;>\r\n<ta...  
2099  <div style=width: 320px; float: right;>\r\n<ta...  
2100  <table style=width: 100%; max-width: 300px; ma...  
2101  <table style=width: 100%; max-width: 300px; ma...

### In order for my program to work, I need each mailing to be in a single row, not multiple rows for every tag. I do this by taking every tag for a single mailing and turning it into a list.  Then I can 'flatten' the current table.  Now, instead of 2101 rows, my table only has 1399 rows, one for each mailing.

In [33]:
flat = pd.DataFrame(camp_txt)  #copy the current dataframe into a new one (I found this helpful for troubleshooting)

def flatten_frame(df,col):  #df= DataFrame, col= the column to be flattened; in our case, 'tag_name'
    df[col] = df[col].fillna('') #if there are no tags, add an empty string
    headers = list(df.columns.values) #pull in the list of columns
    group_cols = list(set(headers) - set([col])) #get the cols to group on by subtracting the one we are flattening
    df = pd.DataFrame(df.groupby(by=(group_cols))[col].apply(list)).reset_index() #group by all the columns except the tags
    df[col] = df[col].apply(lambda x: ', '.join(x)) #convert the flattened col of tags into a string
    return df

flat = flatten_frame(flat,'tag_name') # executes the function defined above
flat = flat[['page_id','tag_name','html']].sort_values(by='page_id')  #sorts the new flattened table
flat = flat.reset_index(drop=True) #needed so the index keeps the same order of the page_ids

pd.options.display.max_colwidth = 110
flat.tail(10)  #displays the last 10 rows of the table

page_id                               tag_name  \
1390    16102                         #Womens_Rights   
1391    16118       taxes, #Economic_Justice, #trade   
1392    16148                           #Environment   
1393    16154                           #Environment   
1394    16160                     bees, #Environment   
1395    16416                           #Environment   
1396    16458                        #Workers_Rights   
1397    16479                        #Workers_Rights   
1398    16736  #Privatization_and_Political_Meddling   
1399    16940              oil company, #Environment   

                                                                                                               html  
1390  <div style=width: 320px; float: right;>\r\n<table style=width: 300px; margin: 10px; border=0 cellspacing=0...  
1391  <div style=width: 320px; float: right;>\r\n<table style=width: 300px; margin: 10px; border=0 cellspacing=0...  
1392  <div style=width: 320px; float: right;>\r\n<table style=width: 300px; margin: 10px; border=0 cellspacing=0...  
1393  <div style=width: 320px; float: right;>\r\n<table style=width: 300px; margin: 10px; border=0 cellspacing=0...  
1394  <div style=width: 320px; float: right;>\r\n<table style=width: 300px; margin: 10px; border=0 cellspacing=0...  
1395  <p>Dear {{ user.first_name|capfirst|default:Friend }}</p>\r\n<p><strong>Water issues are sure heating up a...  
1396  <div style=width: 320px; float: right;>\r\n<table style=width: 300px; margin: 10px; border=0 cellspacing=0...  
1397  <div style=width: 320px; float: right;>\r\n<table style=width: 300px; margin: 10px; border=0 cellspacing=0...  
1398  <div style=width: 320px; float: right;>\r\n<table style=width: 300px; margin: 10px; border=0 cellspacing=0...  
1399  <table style=width: 100%; max-width: 300px; margin-left: 10px; margin-right: 10px; border=0 cellspacing=0 ...

### Now each campaign is on a single row, but in order to analyze the mailing text, I have to extract plain text from all that ugly HTML/Django. I do this with the help of a module called 'Beautiful Soup'. I also have a separate function that takes out all the punctuation from both the html and tag_name fields; I will need that done before the next stage of the analysis.

In [4]:
from bs4 import BeautifulSoup # a module that works with HTML
import re
import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning)

clean = pd.DataFrame(flat) #copy the results from the last step into a new frame

def clean_soup(df,old_col,new_col): # df=DataFrame, col=the column with the dirty HTML we want to clean
    for index, item in df[old_col].iteritems(): #go row by row through the column
        soup = BeautifulSoup(item, "lxml") #turn the current item into a BeautfulSoup object
        washed = soup.get_text(" ",strip=True) #get text from the soup object and store the text in the washed variable
        df.set_value(index,new_col,washed) #update the clean data frame with the washed text
    df[new_col] = df[new_col].str.replace('{(.+)}', ' ') #remove the django tags
    return df

def remove_punc(df,old_col,new_col):
    df[new_col] = df[old_col].str.replace('[^\w\s]',' ') #replaces most punctuation with spaces
    df[new_col] = df[new_col].str.replace('[_]',' ') #replaces underscores with spaces    
    return df

clean['text_clean'] ='' #a new column for storing our squeaky-clean text
clean['tags_clean'] = '' #a new column for storing our squeaky-clean tags
 
clean = clean_soup(clean,'html','text_clean')  #run my function to clean the html
clean = remove_punc(clean,'text_clean','text_clean')  #run my function to remove punctuation from the text
clean = remove_punc(clean,'tag_name','tags_clean') #run my function to premove punctuation from the tags

pd.options.display.max_colwidth = 95
clean[['page_id','html','text_clean']].tail(10)

page_id  \
7       16102   
3       16118   
16      16148   
17      16154   
14      16160   
22      16416   
9       16458   
4       16479   
1       16736   
1026    16940   

                                                                                                html  \
7     <div style=width: 320px; float: right;>\r\n<table style=width: 300px; margin: 10px; border=...   
3     <div style=width: 320px; float: right;>\r\n<table style=width: 300px; margin: 10px; border=...   
16    <div style=width: 320px; float: right;>\r\n<table style=width: 300px; margin: 10px; border=...   
17    <div style=width: 320px; float: right;>\r\n<table style=width: 300px; margin: 10px; border=...   
14    <div style=width: 320px; float: right;>\r\n<table style=width: 300px; margin: 10px; border=...   
22    <p>Dear {{ user.first_name|capfirst|default:Friend }}</p>\r\n<p><strong>Water issues are su...   
9     <div style=width: 320px; float: right;>\r\n<table style=width: 300px; margin: 10px; border=...   
4     <div style=width: 320px; float: right;>\r\n<table style=width: 300px; margin: 10px; border=...   
1     <div style=width: 320px; float: right;>\r\n<table style=width: 300px; margin: 10px; border=...   
1026  <table style=width: 100%; max-width: 300px; margin-left: 10px; margin-right: 10px; border=0...   

                                                                                          text_clean  
7     Craigslist allow exploitative adverts offering homeless women accommodation for sex  Hundre...  
3     Apple is avoiding more than  13 billion in taxes  And that s not even the half of it  Deman...  
16    The world s two biggest greenhouse gas polluters have now ratified the Paris Agreement   so...  
17    The world s two biggest greenhouse gas polluters have now ratified the Paris Agreement   so...  
14    Regulators are rolling out a controversial pesticide banned in Europe to kill Zika mosquito...  
22    Dear   Great news  After 225000 of us spoke out against the mismanagement of the precious w...  
9     Foreign worker lists  That s what this government wants to make all firms publish  Show tha...  
4     Asos is coming under fire for treating its workers  like machines   Tell Asos to treat its ...  
1     Corporations have the government s ear when it comes to Brexit   it promised Nissan tariff ...  
1026  Urgent  The Trudeau government could approve the Kinder Morgan tar sands pipeline in three ...

### Finally we have clear, readable text! However, not all of these words are salient for my purposes; they won't help my algothtm figuring out which campaigns belong to which topics. 

### At this point, I need to start whittling down this text to the words that will be most helpful for our analysis. On thing I consider is the part of speech.  Things like prepositions and conjunctions will confuse the LDA when it tries to dicern discrete topics.  Verbs and adverbs are largely unhelpful too.

### A Python module called the Natural Language Toolkit (NLTK) will help mw filter all the words in the mailing by part of speech.  After some trial and error, I decide to keep only nouns and plural nouns.  I do not inclde proper nouns, which strips out the names of the corporations we target. 

In [5]:
import nltk  # import the Natural Language Toolkit, which includes various tools for text analysis
from nltk.tokenize import sent_tokenize, word_tokenize, PunktSentenceTokenizer
from nltk.corpus import stopwords
from nltk.tokenize import *

filtered = pd.DataFrame(clean[['page_id','text_clean','tags_clean']]) #copy relevant results of last step into a new frame

def filter_camp_by_pos(df,old_col,new_col,pos_codes):
    df[new_col] = df.apply(lambda row: nltk.word_tokenize(row[old_col]), axis=1) #tokenize (pre-process) each word
    df[new_col] = df[new_col].apply(lambda row: nltk.pos_tag(row)) #tag each word with the part of speech
    df[new_col] = [[tuple[0].lower() for tuple in row if tuple[1] in pos_codes] for row in df[new_col]]  #select all words matching my filter codes
    return df

pos_codes = ['NN','NNS']  #just nouns and plural nouns

filter_camp_by_pos(filtered,'text_clean','text_filtered',pos_codes)  #runs the function I defined about

pd.options.display.max_colwidth = 80
filtered[['page_id','text_clean','text_filtered']].tail(10)  #displays last 10 rows

page_id  \
7       16102   
3       16118   
16      16148   
17      16154   
14      16160   
22      16416   
9       16458   
4       16479   
1       16736   
1026    16940   

                                                                           text_clean  \
7     Craigslist allow exploitative adverts offering homeless women accommodation ...   
3     Apple is avoiding more than  13 billion in taxes  And that s not even the ha...   
16    The world s two biggest greenhouse gas polluters have now ratified the Paris...   
17    The world s two biggest greenhouse gas polluters have now ratified the Paris...   
14    Regulators are rolling out a controversial pesticide banned in Europe to kil...   
22    Dear   Great news  After 225000 of us spoke out against the mismanagement of...   
9     Foreign worker lists  That s what this government wants to make all firms pu...   
4     Asos is coming under fire for treating its workers  like machines   Tell Aso...   
1     Corporations have the government s ear when it comes to Brexit   it promised...   
1026  Urgent  The Trudeau government could approve the Kinder Morgan tar sands pip...   

                                                                        text_filtered  
7     [adverts, women, accommodation, sex, hundreds, vile, ads, petition, bans, pe...  
3     [taxes, half, tax, evasion, petition, secret, technology, giant, profits, go...  
16    [world, s, greenhouse, gas, polluters, excuse, fossil, fuel, industry, clima...  
17    [world, s, greenhouse, gas, polluters, governments, excuse, governments, fos...  
14    [regulators, pesticide, mosquitoes, bees, bees, regulators, spraying, petiti...  
22    [news, mismanagement, water, water, rates, users, companies, milestone, towa...  
9     [worker, lists, government, firms, plans, workers, workers, headline, days, ...  
4     [fire, workers, machines, workers, petition, cost, fashion, giant, clothes, ...  
1     [corporations, government, tariff, access, market, corporations, brexit, dea...  
1026  [government, sands, pipeline, days, dozens, organizations, hundreds, thousan...

### Up until this point, I have kept the text of mailings seperate from the tags, but at this point I want to grab the words from the 'tag_name' column and add them to the mailing text to be evaluated together.  I couldn't do that until after I ran the part of speech filter, because NLTK needs to have sentences in context in order to  properly mark the part of speech; adding short phrases would have created problems.  From here on in, I will be dealing with collections of words, where the order doesn't matter.  Because the tags were specifically chosen to convey topic information, I have chosen to give them 2x more weight than the regular mailing text, based on trial and error.

In [6]:
merged = pd.DataFrame(filtered[['page_id','text_clean','text_filtered','tags_clean']]) #copy results of last step into a new frame

def concat_cols(df, filtered_col, unfiltered_col, new_col, coef):
    weighted = ((df[unfiltered_col].str.lower()+' ')*coef) #multiply the unfiltered column by the desired coefficient
    df[new_col] = weighted.apply(lambda row: nltk.word_tokenize(row)) #tokenize (pre-process) each word 
    df[new_col] = df[new_col] + df[filtered_col]
    return df

merged['text_merged'] = '' # a new column to store the combined result of text and tags                     

merged = concat_cols(merged,'text_filtered','tags_clean','text_merged',3)

pd.options.display.max_colwidth =70
merged[['page_id','text_filtered','tags_clean','text_merged']].tail(100)

page_id  \
1093    11061   
1033    11063   
1230    11064   
1228    11078   
1058    11085   
1005    11086   
1084    11087   
1085    11090   
1021    11096   
1059    11101   
1070    11157   
1099    11165   
1015    11169   
1110    11202   
1071    11212   
1009    11214   
1081    11227   
1101    11232   
1016    11235   
1111    11316   
1067    11323   
1038    11328   
1047    11329   
1006    11343   
1089    11357   
1122    11360   
1069    11361   
1090    11367   
1098    11376   
1034    11416   
...       ...   
1116    12671   
1117    12672   
1100    12689   
1105    12705   
1037    12708   
1041    12717   
1074    12781   
1112    12789   
1002    12790   
15      12912   
19      14030   
5       14483   
6       14728   
10      14887   
8       15109   
2       15316   
12      15669   
13      15795   
11      15815   
18      15966   
7       16102   
3       16118   
16      16148   
17      16154   
14      16160   
22      16416   
9       16458   
4       16479   
1       16736   
1026    16940   

                                                              text_filtered  \
1093  [people, taxes, people, petition, people, tax, deal, billions, hea...   
1033  [report, executives, toymakers, kind, message, fans, girls, heroes...   
1230  [report, women, men, products, culprits, practice, call, women, pr...   
1228  [listeria, outbreak, hospital, salad, processing, plant, response,...   
1058  [hundreds, millions, year, taxes, corporations, taxes, petition, h...   
1005  [days, law, call, government, power, grab, petition, s, trade, dea...   
1084  [nobody, fees, banks, banks, year, fees, fees, racket, time, s, ba...   
1085  [government, share, taxes, tax, system, petition, government, shar...   
1021  [week, corporations, power, ratification, petition, days, things, ...   
1059  [grocer, giant, produce, bee, pesticides, shops, compatriot, super...   
1070  [victory, workers, judge, employees, attitude, workers, rights, re...   
1099  [tax, deal, somebody, complaint, letter, complaint, tax, name, tax...   
1015  [s, carcinogen, s, shelves, product, petition, year, ingredient, w...   
1110  [acts, disobedience, generation, democracy, end, influence, electi...   
1071  [midst, trial, damages, party, contribution, party, money, hold, p...   
1009  [placebo, drug, company, fire, practices, product, work, petition,...   
1081  [children, water, end, s, operations, people, water, petition, wat...   
1101  [draft, bill, mass, surveillance, petition, lives, risk, deal, gia...   
1016  [waterways, mess, petition, foes, city, suit, agribusiness, giant,...   
1111  [s, chocolate, slavery, marks, end, child, labour, chocolate, indu...   
1067  [executive, job, door, government, tax, dodgers, petition, boss, j...   
1038  [thousands, iphones, pieces, junk, software, iphones, parties, pet...   
1047  [car, emissions, tests, manufacturers, chance, car, industry, peti...   
1006  [report, conflict, minerals, electronics, market, exploitation, su...   
1089  [public, ads, station, dodgy, ads, petition, public, ads, station,...   
1122  [corporations, planet, s, resources, oceans, litter, economy, poli...   
1069  [pounds, pollution, waterways, water, pollution, policy, petition,...   
1090  [executive, order, donations, corporations, business, president, s...   
1098  [restrictions, donations, lobbyists, restrictions, touch, billiona...   
1034  [articles, monopoly, publishing, research, s, fight, publisher, la...   
...                                                                     ...   
1116  [meeting, company, resolution, workers, rights, workers, name, sha...   
1117  [meeting, company, resolution, workers, rights, workers, name, sha...   
1100  [government, coal, seam, gas, mining, interests, right, government...   
1105  [government, school, money, hands, corporations, don, t, schools, ...   
1037  [iphone, headphone, jack, e, waste, move, time, iphone, headphone,...   
1041  [bosses, bill, comment

### Next I need to assemble a list of words that I want to exclude from the analysis. NLTK has gotten rid of many of the words we don't need based on parts of speech, but other words are specific to SumOfUs use, and have to be listed manually.  Words like 'petition', 'corporation', and 'click' are not salient for us, even though they could be salient in another context. I started with a list of generic stopwords and then added to it by hand.  Using the frequency distribution module of NLTK to make it easier, I can see the most commonly used words in my dataset, and then add the ones I want to exclude to my stoplist.  I ran this step several times until I was left with a list of salient terms.

In [7]:
from nltk.probability import FreqDist

go = pd.DataFrame(merged[['page_id','text_clean','text_merged']]) #copy results of last step into a new frame

def exclude_stopwords(df,old_col,new_col,num):
    stop_words = set(stopwords.words('english')) #read in my text file of stopwords   
    df[new_col] = df[old_col].apply(lambda x:[word for word in x if word not in stop_words]) #remove the stop words
    go_freq = FreqDist(df[new_col].sum()).most_common(num) #return the top words that are left by frequency
    return df, go_freq


go['text_go'] = ''
    
go, go_freq = exclude_stopwords(go,'text_merged','text_go',500)

go_freq #I use this list to find common words that I don't want in my analysis, then manually add them the stoplist

go_freq

[('oil', 1179),
 ('food', 974),
 ('water', 788),
 ('tax', 678),
 ('protection', 665),
 ('climate', 628),
 ('health', 521),
 ('liberties', 512),
 ('trade', 498),
 ('media', 469),
 ('economic', 459),
 ('privatization', 457),
 ('meddling', 438),
 ('women', 427),
 ('gmos', 372),
 ('customers', 355),
 ('children', 310),
 ('coal', 292),
 ('conditions', 283),
 ('farmers', 283),
 ('taxes', 273),
 ('mining', 267),
 ('pipeline', 265),
 ('energy', 263),
 ('bees', 260),
 ('gas', 248),
 ('cancer', 247),
 ('laws', 247),
 ('safety', 237),
 ('accountibility', 222),
 ('palm', 206),
 ('governments', 203),
 ('drug', 197),
 ('womens', 196),
 ('wages', 193),
 ('shareholders', 187),
 ('chemicals', 185),
 ('fracking', 174),
 ('violence', 168),
 ('factory', 167),
 ('fuel', 164),
 ('neonics', 162),
 ('land', 157),
 ('banks', 156),
 ('internet', 156),
 ('pesticides', 156),
 ('lobbying', 154),
 ('prices', 153),
 ('prison', 148),
 ('jobs', 136),
 ('investment', 136),
 ('contract', 130),
 ('treatment', 129),
 ('ta

### Some words have the same root, such as 'work', 'working', and 'worker' or 'economy', 'economic' and 'economists.  For our purposes, these words have basically the same meaning, and should be combined in order to properly represent the weight of each term in the campaign.  I apply three different 'stemmer' algorithms, in order to use the roots of words in my model.

In [8]:
# stem the filtered tokens
from nltk.stem.snowball import SnowballStemmer
from nltk.stem.lancaster import LancasterStemmer
from nltk.stem.porter import PorterStemmer

stemmed = pd.DataFrame(go[['page_id','text_clean','text_go']]) #copy results of last step into a new frame

snowball = SnowballStemmer('english')
lancaster = LancasterStemmer()
porter = PorterStemmer()

def stem(df,old_col,new_col):
    df[new_col] = df[old_col].apply(lambda x: [porter.stem(word) for word in x])
    df[new_col] = df[new_col].apply(lambda x: [snowball.stem(word) for word in x])
    df[new_col] = df[new_col].apply(lambda x: [lancaster.stem(word) for word in x])
    return df

stemmed['text_stemmed'] = ''
stemmed = stem(stemmed,'text_go','text_stemmed')

pd.options.display.max_colwidth =90
stemmed[['page_id','text_go','text_stemmed']].tail(10)

page_id  \
7       16102   
3       16118   
16      16148   
17      16154   
14      16160   
22      16416   
9       16458   
4       16479   
1       16736   
1026    16940   

                                                                                        text_go  \
7     [womens, womens, womens, adverts, women, accommodation, sex, vile, bans, adverts, wome...   
3     [taxes, economic, trade, taxes, economic, trade, taxes, economic, trade, taxes, tax, e...   
16    [greenhouse, gas, polluters, excuse, fuel, climate, climate, treaty, force, treaty, fu...   
17    [greenhouse, gas, polluters, governments, excuse, governments, fuel, climate, governme...   
14    [bees, bees, bees, regulators, pesticide, mosquitoes, bees, bees, regulators, spraying...   
22    [mismanagement, water, water, rates, milestone, water, ecology, economy, conversation,...   
9     [lists, firms, headline, flames, racism, businesses, workforce, lists, doctors, visa, ...   
4     [machines, fashion, clothes, expense, warehouse, warehouse, water, toilet, breaks, per...   
1     [privatization, meddling, privatization, meddling, privatization, meddling, tariff, br...   
1026  [oil, oil, oil, pipeline, tar, pipeline, megaproject, messages, pipeline, pipeline, co...   

                                                                                   text_stemmed  
7     [wom, wom, wom, advert, wom, accommod, sex, vil, ban, advert, wom, exchang, sex, wom, ...  
3     [tax, econom, trad, tax, econom, trad, tax, econom, trad, tax, tax, ev, secret, techno...  
16    [greenh, ga, pollut, exc, fuel, clim, clim, treat, forc, treat, fuel, clim, forc, clim...  
17    [greenh, ga, pollut, govern, exc, govern, fuel, clim, govern, govern, treat, fuel, cli...  
14    [bee, bee, bee, reg, pesticid, mosquito, bee, bee, reg, spray, mosquito, reg, pesticid...  
22    [mism, wat, wat, rat, mileston, wat, ecolog, econom, conv, anch, med, wat, approv, com...  
9     [list, firm, headlin, flam, rac, bus, workforc, list, doct, vis, stud, workforc, badg,...  
4     [machin, fash, clo, exp, wareh, wareh, wat, toilet, break, perform, second, turnov, il...  
1     [priv, meddl, priv, meddl, priv, meddl, tariff, brexit, strategi, min, car, auto, nego...  
1026  [oil, oil, oil, pipelin, tar, pipelin, megaproject, mess, pipelin, pipelin, controvers...

### Now, at long last, I am finally able to get to the good stuff and build my model!  I use a module called gensim; the name is derived from “generate similar” because it generates topics based on the similarity of words within each group.  The first step is to take my painstakingly cleaned and filtered text and create a giant "bag of words", called a corpus, to feed into a Latent Dirichlet Allocation (LDA) model.  The model will derive topics from our corpus based on word frequency.   It then returns the top four words by frequency in each topic.  The coefficients represent the percentage of that word in the topic as a whole.  This alone does not tell you that much about the topics, so in the next step, I generate an interactive graphic to explore the topics in more depth.

In [9]:
from gensim import corpora, models

def generate_lda(df,col,topics,words,num_passes):
    camp_corpus = df[col].tolist()
    dictionary = corpora.Dictionary(camp_corpus)
    corpus = [dictionary.doc2bow(text) for text in camp_corpus]
    lda = models.ldamodel.LdaModel(corpus, num_topics=topics, id2word=dictionary, passes=num_passes)
    topics = lda.print_topics(num_topics=topics,num_words=words)
    return lda, corpus, dictionary, topics

lda, corpus, dictionary, all_topics = generate_lda(stemmed,'text_stemmed', 8, 4, 150)

all_topics

[(0, '0.027*libert + 0.023*med + 0.023*protect + 0.020*prison'),
 (1, '0.076*wat + 0.046*priv + 0.044*meddl + 0.036*min'),
 (2, '0.122*oil + 0.037*palm + 0.028*forest + 0.021*deforest'),
 (3, '0.052*oil + 0.033*pipelin + 0.019*spil + 0.018*whal'),
 (4, '0.090*tax + 0.043*econom + 0.021*factor + 0.018*wag'),
 (5, '0.091*wom + 0.019*sex + 0.017*med + 0.013*viol'),
 (6, '0.083*food + 0.042*farm + 0.035*gmo + 0.034*bee'),
 (7, '0.048*trad + 0.048*clim + 0.024*coal + 0.024*bank')]

### Here I have generated an interactive visualization that will help me explore and understand each topic.  Each bubble represents a single topic. The size of the bubble indicating the relative proportion of 'tokens' (words) in the corpus that are related to that topic.  The bubbles are numbered by size in descending order and do not match the topic listing in the previous step.

### The topics are placed on a field with two axes: PC1 and PC2.  PC stands for Principal Component. Principal component analysis (PCA) is a technique used to emphasize variation and bring out strong patterns in a dataset. It's often used to make data easy to explore and visualize. The axes don't mean any one thing but represent a composite of features chosen algorithmically to maximize variation.  Which is to say, the placement of the bubbles on the field shows you how closely related the topics are to each other in terms of the words they contain.  

### A single word can be found in more than one topic when it used in different contexts.  For example, the word 'oil' often appears with 'pipeline', and also often appears with 'palm'.  The algorithm makes guesses as to what topic a word belong to, based on other words in the same campaigns.  When two topics share many terms they are shown on the field as overlapping.

###  The blue bars on the right represent the overall frequency of each term in the corpus.  If you mouse over any bubble, red bars appear to the right.  The red bars represent the estimated number of times a given term was included in a given topic.   If you mouse over any word on the right, the bubbles change size to show the be proportional to the frequency of that term.

In [24]:
import pyLDAvis.gensim
import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning)

vis = pyLDAvis.gensim.prepare(lda, corpus, dictionary)
pyLDAvis.display(vis)

### Now that the topics have been generated, I need to apply them back to the campaigns.  The LDA algorithm looks at each campaign, and guesses a percentage chance that campaign is related to each of the topics.  Note that the topic numbers here related to the original numbered list, not the numbers on the bubbles in the visualization. 

# THIS PART NOT WORKING AT ALL YET - COL HEADERS ARE NOT STAYING WITH THE RIGHT NUMBERS FROM THE TUPLES

In [11]:
assigned = pd.DataFrame(stemmed[['page_id','text_clean','text_stemmed']])  #copy results of last step into a new frame

def calc_freq(df,old_col,new_col,num):
    for index, row in df.iterrows():
        camp_freq = df['text_stemmed'][index]
        dist = FreqDist(camp_freq)
        df.set_value(index,'freq',dist.most_common(num))
    return df

def assign_topics(df,old_col,new_col,num):
    for index, row in df.iterrows():
        seed = df[old_col][index]
        doc_bow = dictionary.doc2bow(seed)
        camp_tops = lda.get_document_topics(doc_bow,num)
        df.set_value(index,new_col, camp_tops)
    return df
                              
assigned['freq'] = ''
assigned['topics'] = ''

assigned = calc_freq(assigned,'text_stemmed','freq',20)
assigned = assign_topics(assigned,'text_stemmed','topics',0)

pd.options.display.max_colwidth =150
assigned[['page_id','freq','topics']]

page_id  \
44        824   
45        830   
49        831   
37        834   
29        840   
33        841   
31        876   
36        883   
30        886   
42        891   
40        898   
39        924   
43        935   
47        936   
32        971   
50        973   
41        982   
38        997   
46       1001   
35       1066   
48       1067   
21       1077   
34       1082   
27       1105   
556      1154   
980      1164   
793      1207   
283      1227   
992      1232   
525      1242   
...       ...   
1116    12671   
1117    12672   
1100    12689   
1105    12705   
1037    12708   
1041    12717   
1074    12781   
1112    12789   
1002    12790   
15      12912   
19      14030   
5       14483   
6       14728   
10      14887   
8       15109   
2       15316   
12      15669   
13      15795   
11      15815   
18      15966   
7       16102   
3       16118   
16      16148   
17      16154   
14      16160   
22      16416   
9       16458   
4       16479   
1       16736   
1026    16940   

                                                                                                                                                       freq  \
44    [(priz, 5), (food, 4), (gmo, 3), (agricult, 3), (farm, 3), (cred, 2), (model, 2), (crop, 2), (kil, 2), (ruthless, 1), (mod, 1), (bee, 1), (biotech...   
45    [(edit, 4), (own, 3), (account, 3), (med, 3), (disco, 2), (pap, 2), (legac, 2), (attack, 2), (facad, 1), (oppon, 1), (apolog, 1), (column, 1), (ed...   
49    [(land, 16), (sug, 10), (farm, 7), (bev, 4), (econom, 3), (libert, 3), (produc, 3), (supply, 3), (grab, 2), (tol, 2), (hom, 2), (commod, 1), (inti...   
37    [(steel, 4), (hom, 4), (econom, 3), (vil, 3), (expert, 2), (invest, 2), (plant, 2), (eth, 1), (wav, 1), (reput, 1), (lifestyl, 1), (med, 1), (subs...   
29    [(factor, 9), (disast, 4), (supply, 3), (clo, 3), (med, 2), (safet, 2), (tol, 2), (repair, 1), (renov, 1), (shirt, 1), (break, 1), (inspect, 1), (...   
33    [(factor, 12), (safet, 9), (inspect, 3), (bind, 3), (traged, 3), (yesterday, 2), (disast, 2), (apparel, 2), (comp, 2), (med, 2), (collap, 2), (rep...   
31    [(salmon, 6), (gold, 5), (min, 4), (gen, 2), (liv, 2), (mil, 1), (sockey, 1), (stream, 1), (pristin, 1), (bay, 1), (land, 1), (gree, 1), (tabl, 1)...   
36    [(bank, 10), (priv, 3), (spec, 3), (meddl, 3), (op, 3), (fund, 3), (hedg, 2), (valu, 2), (stock, 2), (principl, 2), (custom, 2), (rescu, 2), (inde...   
30    [(wom, 12), (wit, 6), (libert, 3), (rap, 3), (contract, 3), (det, 3), (pattern, 2), (surv, 2), (seek, 2), (immigr, 2), (remov, 2), (prison, 1), (s...   
42    [(yard, 5), (rail, 4), (econom, 3), (job, 3), (wareh, 3), (strike, 3), (weal, 2), (wag, 2), (inequ, 2), (icon, 1), (extrem, 1), (povert, 1), (resp...   
40    [(pipelin, 14), (premy, 8), (oil, 4), (clim, 4), (annount, 3), (provint, 3), (tar, 3), (habit, 2), (surpr, 2), (condit, 2), (riv, 1), (med, 1), (a...   
39    [(coal, 5), (express, 3), (reef, 3), (clim, 3), (barry, 3), (build, 2), (sustain, 2), (reput, 2), (icon, 2), (port, 2), (grav, 1), (doorstep, 1), ...   
43    [(whal, 13), (plast, 9), (supermarket, 6), (wast, 5), (supply, 4), (greenh, 4), (protect, 3), (oc, 3), (environ, 3), (fruit, 2), (veget, 2), (piec...   
47    [(flight, 7), (dog, 7), (passeng, 7), (seat, 4), (anim, 4), (libert, 3), (plan, 3), (airlin, 2), (respect, 2), (attend, 2), (delay, 2), (crew, 2),...   
32    [(food, 6), (art, 5), (bull, 5), (fil, 4), (gmo, 3), (trademark, 3), (coff, 3), (kal, 3), (chick, 2), (slog, 2), (guy, 2), (pat, 2), (town, 2), (r...   
50    [(bee, 18), (pesticid, 6), (food, 4), (neon, 3), (gmo, 3), (honeyb, 2), (ecosystem, 2), (pollin, 2), (chem, 2), (farm, 2), (gard, 2), (restrict, 2...   
41    [(trad, 12), (govern, 9), (law, 4), (transatl, 3), (pact, 3), (agr, 3), (ttip, 3), (invest, 3), (democrac, 3), (court, 2), (treat, 2), (heal, 2), ...   
38    [(produc, 7), (rabbit, 6), (angor, 6), (anim, 5), (retail, 4), (farm, 3), (clo, 3),

### Now I need to break down that topics field so that each topic gets assigned to its own column. For readibility, I will assign each column a name, and convert the numbers into rounded percentages. Finally, I will both display the final table and save it as a csv for use in my linear regression.  The linear regresssion model will determine how predictive each topic is of a cohort's propensity to donate.  These topics will be combined with other information from the campaign like regional breakdown, campaign virality, and early behavior of new joiners from that cohort.  All of the information will be used to determine the best way to predict the probability that a new member from a given campaign will become a donor.

In [23]:
pivoted = pd.DataFrame(assigned[['page_id','text_clean','freq','topics']]) 
                                  
def pivot(df,topic_col):
    topicframe = pd.DataFrame()
    topicframe[topic_col] = [[tuple[1] for tuple in topic] for topic in df[topic_col]]  #removes a layer of nesting
    topicframe = pd.DataFrame(topicframe[topic_col].tolist())
    topicframe = round(topicframe*100,0)
    #topicframe.columns=headers
    
    df = pd.concat([df, topicframe], axis=1)
    return df

topic_headers = ['0','1','2','3','4','5','6','7']

#topic_headers = ['health','private','palm','fossil','econ','rights','food','trade']

pivoted = pivot(pivoted,'topics')

pivoted.to_csv('mailing_topic.csv')

pd.options.display.max_colwidth =70
#pivoted[['page_id','topics']+topic_headers]
pivoted

page_id  \
0        3180   
1       16736   
2       15316   
3       16118   
4       16479   
5       14483   
6       14728   
7       16102   
8       15109   
9       16458   
10      14887   
11      15815   
12      15669   
13      15795   
14      16160   
15      12912   
16      16148   
17      16154   
18      15966   
19      14030   
20       3376   
21       1077   
22      16416   
23       1777   
24      12544   
25      11026   
26       2634   
27       1105   
28       1258   
29        840   
...       ...   
1370    10633   
1371     8564   
1372    10376   
1373     8848   
1374     4894   
1375    10531   
1376     6430   
1377    10439   
1378     7412   
1379     8462   
1380     8571   
1381     8575   
1382    10511   
1383    10663   
1384     6506   
1385     7406   
1386     8918   
1387     8661   
1388     5259   
1389    10653   
1390    10406   
1391    10666   
1392    10565   
1393     6363   
1394    10472   
1395     6485   
1396     6489   
1397    10764   
1398     3230   
1399     3529   

                                                                 text_clean  \
0       This year s UN climate talks in Paris will be a pivotal moment t...   
1     Corporations have the government s ear when it comes to Brexit   i...   
2     Grassy Narrows community members continue to report on going healt...   
3     Apple is avoiding more than  13 billion in taxes  And that s not e...   
4     Asos is coming under fire for treating its workers  like machines ...   
5     BREAKING  Malcolm Turnbull has just been outed in the Panama Paper...   
6     Bayer s offered to buy Monsanto creating a gigantic mega corporati...   
7     Craigslist allow exploitative adverts offering homeless women acco...   
8     Despite fierce opposition Glyphosate is about to get another stamp...   
9     Foreign worker lists  That s what this government wants to make al...   
10    Fracking could go ahead in the UK for the first time in five years...   
11    Jason Clare has just been appointed the new Shadow Trade Minister ...   
12    Nestlé wants to build a massive water bottling plant in Phoenix Ar...   
13    Next week PepsiCo is re launching Crystal Pepsi  While it s in the...   
14    Regulators are rolling out a controversial pesticide banned in Eur...   
15    The Panama Papers leak makes it crystal clear that when it comes t...   
16    The world s two biggest greenhouse gas polluters have now ratified...   
17    The world s two biggest greenhouse gas polluters have now ratified...   
18    Unfinished bathrooms  Green pools  Exposed wires  Clogged toilets ...   
19    While average workers struggle to get by CEOs take home outrageous...   
20      Nearly 50000 of you have signed the petition to halt Chris Chris...   
21      A global construction firm wants to build a massive coal port on...   
22    Dear   Great news  After 225000 of us spoke out against the misman...   
23    Dear    Torture   That s how Senator John McCain described his own...   
24    Hey   I m Emma and I m running our WestJet campaign that has been ...   
25      The government has nearly succeeded in pushing through its anti ...   
26      World leaders are racing against time to push through two trade ...   
27      Tony Abbott wants to strip Tasmania s forests of their World Her...   
28      Stephen Harper s trade minister is about to sign off on the worl...   
29    A deadly factory fire has just broken out in The Bay s supplier fa...   
...                                                                     ...   
1370  There s a new study out on neonics    and it s bad news for apples...   
1371  This could be it  world leaders are gathering in Atlanta this week...   
1372  Tim Hortons and Starbucks say they recycle paper coffee cups    bu...   
1373  Tim Laidler is a star candidate for the Conservatives    but his c...   
1374  Tony Abbot promised he d leave the ABC alone then he cut  254 mill...   
1375  Total  claims to suppo